flang/lib/Optimizer/Transforms/AffinePromotion.cpp

```diff
@@ -100,37 +189,97 @@ class AffineLoopConversion : public mlir::OpRewritePattern<fir::LoopOp> {
         AffineMap::getMultiDimIdentityMap(1, loop.getContext()),
         ValueRange(loop.upperBound()),
         AffineMap::getMultiDimIdentityMap(1, loop.getContext()),
         loopAnalysis.step.getValue());
-    loop.step().dropAllUses();
-    loop.getBody()->getOperations().pop_back(); // remove fir.result
+
+    rewriter.startRootUpdate(affineFor.getOperation());
     affineFor.getBody()->getOperations().splice(affineFor.getBody()->begin(),
-                                                *loopOps, loopOps->begin(),
-                                                loopOps->end());
-    rewriter.eraseOp(loop);
+                                                loopOps, loopOps.begin(),
+                                                --loopOps.end());
+    rewriter.finalizeRootUpdate(affineFor.getOperation());
+
+    for (auto &bodyOp : affineFor.getBody()->getOperations()) {
+      if (isa<fir::LoadOp>(bodyOp)) {
+        if (failed(rewriteLoad(cast<fir::LoadOp>(bodyOp), rewriter))) {
+          return failure();
+        }
+      }
+      if (isa<fir::StoreOp>(bodyOp)) {
+        if (failed(rewriteStore(cast<fir::StoreOp>(bodyOp), rewriter))) {
+          return failure();
+        }
+      }
+    }
+
+    rewriter.startRootUpdate(loop.getOperation());
     loop.getInductionVar().replaceAllUsesWith(affineFor.getInductionVar());
+    rewriter.finalizeRootUpdate(loop.getOperation());
+
+    rewriter.replaceOp(loop, affineFor.getOperation()->getResults());
+
+    LLVM_DEBUG(llvm::dbgs() << "AffineLoopConversion: loop rewriten to:\n";
+               affineFor.dump(););
     return success();
   }
 
 private:
-  mlir::LogicalResult rewriteLoad(fir::LoadOp op,
+  mlir::Type coordinateArrayElement(fir::ArrayCoorOp op) const {
+    if (auto refType = op.ref().getType().dyn_cast_or_null<ReferenceType>()) {
+      if (auto seqType = refType.getEleTy().dyn_cast_or_null<SequenceType>()) {
+        return seqType.getEleTy();
+      }
+    }
+    op.emitError(
+        "AffineLoopConversion: array type in coordinate operation not valid\n");
+  }
+  std::pair<mlir::AffineApplyOp, fir::ConvertOp>
+  createAffineOps(mlir::Value arrayRef, mlir::PatternRewriter &rewriter) const {
+    auto acoOp = arrayRef.getDefiningOp<ArrayCoorOp>();
+    auto genDim = acoOp.dims().getDefiningOp<GenDimsOp>();
+    auto affineMap =
+        createArrayIndexAffineMap(acoOp.coor().size(), acoOp.getContext());
+    SmallVector<mlir::Value, 4> indexArgs;
+    indexArgs.append(acoOp.coor().begin(), acoOp.coor().end());
+    indexArgs.append(genDim.triples().begin(), genDim.triples().end());
+    auto affineApply = rewriter.create<mlir::AffineApplyOp>(
+        acoOp.getLoc(), affineMap, indexArgs);
+    auto arrayElementType = coordinateArrayElement(acoOp);
+    auto newType = mlir::MemRefType::get({-1}, arrayElementType);
+    auto arrayConvert =
+        rewriter.create<fir::ConvertOp>(acoOp.getLoc(), newType, acoOp.ref());
```

Comment:
Is this cast from `!fir.ref<!fir.array<?x?xf32>>` to  `memref<?xf32>` legal ? I thought `memref` was a rich pointer  (dynamic extents stored into it, like a descriptor) and there is no such dynamic info in `!fir.ref<!fir.array<?x?xf32>> that is just a pointer to the storage.

Reply:
Affine dialect has a hard dependency on `memref` so for now the plan is to add a convert and remove after using affine optimizations.